File: math_junior.c
#include <stdio.h>

#define SIZE 4

int main();
int solve_cell();
int is_valid(int row, int column, int value);
void print_grid();

int grid[5][5] = {
        { 1, 0, 0,12,31},
		{ 0, 0, 0, 0,27},
		{ 0, 0, 0, 0,29},
		{16, 0, 0,10,49},
		{43,37,23,33, 0} };

int main()
{
    solve_cell();
	return 0;
}

int solve_cell()
{
	for (int i = 0; i < SIZE; i++)
    {
        for (int j = 0; j < SIZE; j++)
        {
            /* Find blank cell */
            if (grid[i][j] == 0)
            {
                /* Loop from 1 - 16 */
                for (int n = 1; n < 17; n++)
                {
                    if (is_valid(i, j, n))
                    {
                        grid[i][j] = n;
                        solve_cell();

                        grid[i][j] = 0;
                    }
                    
                }
                return 0;
            }
        }
    }
    static int count;
    printf("id:%d\n", ++count);
    print_grid();
}

int is_valid(int row, int col, int num)
{
    int count;
    // can't repeat
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            if (grid[i][j] == num) return 0;            
        }
        
    }
    
    // Check row
    count = 0;
    for (int j = 0; j < SIZE; j++)
        count += grid[row][j];
    count += num;
    if (count > grid[row][4])
        return 0;
    int with_value = 1;
    for (int j = 0; j < SIZE; j++) {
        if (j != col)
            if (grid[row][j] == 0) with_value = 0;
    }
    if (with_value) 
        if (count != grid[row][4]) return 0;

    // Check column
    count = 0;
    for (int i = 0; i < SIZE; i++)
        count += grid[i][col];
    count += num;
    if (count > grid[4][col])
        return 0;
    with_value = 1;
    for (int i = 0; i < SIZE; i++) {
        if (i != row)
            if (grid[i][col] == 0) with_value = 0;
    }
    if (with_value)
        if (count != grid[4][col]) return 0;

    return 1;
}

void print_grid()
{
	int i, j;

	for (i = 0; i < 4; i++) {
		for (j = 0; j < 4; j++) {
			grid[i][j] != 0 ? printf("%2d ", grid[i][j]) : printf(" . ");
		}
		printf("\n");
	}
}

File: math_junior.py
import numpy as np

count = 0
# the black cells are filled with 0
grid = [[ 1, 0, 0,12,31],
        [ 0, 0, 0, 0,27],
        [ 0, 0, 0, 0,29],
        [16, 0, 0,10,49],
        [43,37,23,33, 0]]

def possible(y,x,n):
  global grid
  # n is the number we want to fill in

  # 3rd
  # can't repeat
  for j in range(4):
    for i in range(4):
      if grid[j][i] == n:
        return False

  # 1st , check row y
  count = 0
  for i in range(4):
    count += grid[y][i]
  count += n
  if count > grid[y][4]:
    return False
  with_value = True
  for i in range(4):
    if i != x:
      if grid[y][i] == 0:
        with_value = False
  if with_value:
    if count != grid[y][4]:
      return False
  # check row y, if all 4 cell are bigger than 0
  #if grid[y][0] > 0 and grid[y][1] > 0 and grid[y][2] > 0 and grid[y][3] > 0:
  #  if count != grid[y][4]:
  #    return False

  # 2nd, check column x
  count = 0
  for i in range(4):
    count += grid[i][x]
  count += n
  if count > grid[4][x]:
    return False
  with_value = True
  for j in range(4):
    if j != y:
      if grid[j][x] == 0:
        with_value = False
  if with_value:
    if count != grid[4][x]:
        return False
  # check column x, if all 4 cell are bigger than 0
  #if grid[0][x] > 0 and grid[1][x] > 0 and grid[2][x] > 0 and grid[3][x] > 0:
  #  if count != grid[4][x]:
  #    return False

  # return true if pass all 2 checks.
  return True

def solve():
  global grid, count
  for y in range(4):
    for x in range(4):
      # Find blank positions in the grid (value = 0)
      if grid[y][x] == 0:
        # Loop n from 1-16
        for n in range(1,17):
          # print("{},{},{}".format(y,x,n))
          if possible(y,x,n):
            grid[y][x] = n
            solve()

            # This is where backtracking happens
            # Reset the latest position back to 0 and try with new n value
            grid[y][x] = 0
        return
  count = count + 1
  print("id:{}".format(count))
  print(np.matrix(grid))
  #input('More?')

solve()
